### DISCUSSION TITLE
Parsing Message for Multiple ZAD segments and removing them based on what information is present

### DISCUSSION THREAD

**Discussion by dombarone** [nextgenhealthcare/connect#6006] [Q&A] [answered]
Good afternoon, 

I'm looking to have a transformer in Mirth remove a ZAD segment if ZAD.1 == 0 and the rest of the segments are blank. I can get this to work when there is only one FT1 segment, but we send out charges with multiple FT1 segments and the logic doesn't work when there are multiple. 

**Sample Message** 
MSH|^~\&|RCx|10020308|Ntierprise|Michigan Institute Of Urology PC|20231201095330||DFT^P03|04ce5874-429e-4fd7-813c-5421b7679223|P|2.7|||AL
EVN|P03|20231201095330334
PID|1|6326310|6326310||Test^Milo||20101110|M|||Home lane ^^Gretna^LA^70056||||||||567-89-0432|||||||||||N
PV1|1|O|UCBR|||||Rubin||||||||||||||||||||||||||||||||||||||||||11014907
FT1|1|48965089-d14c-40f5-adf7-75b7196dde80|1025|20231121^||CG|99213|||1|||UC|||UCBR|||V80.1^^I10~W56.51^^I10|JA01Anem|JA01Anem||||99213|||||||
ZVA|48965089-d14c-40f5-adf7-75b7196dde80
ZDS|||||||
ZAD|0|||||||||||||||||||||||||||||||||||||||
FT1|2|858a6556-074c-4e54-90ca-9068fb115bcc|1025|20231121^||CG|80050|||1|||UC|||UCBR|||V80.1^^I10~W56.51^^I10|JA01Anem|JA01Anem||||80050|||||||
ZVA|858a6556-074c-4e54-90ca-9068fb115bcc
ZDS|||||||
ZAD|0|||||||||||||||||||||||||||||||||||||||

**Current iteration of transformer** 
```javascript
var count=0;
    for each (segment in msg.children() ){
        if(segment.name() == "ZAD"){
                if ((msg['ZAD'][count]['ZAD.1']['ZAD.1.1'].toString() == "0") 
                && (msg['ZAD'][count]['ZAD.2']['ZAD.2.1'].toString() == "") 
                && (msg['ZAD'][count]['ZAD.3']['ZAD.3.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.4']['ZAD.4.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.5']['ZAD.5.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.6']['ZAD.6.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.7']['ZAD.7.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.8']['ZAD.8.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.9']['ZAD.9.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.10']['ZAD.10.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.11']['ZAD.11.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.12']['ZAD.12.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.13']['ZAD.13.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.14']['ZAD.14.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.15']['ZAD.15.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.16']['ZAD.16.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.17']['ZAD.17.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.18']['ZAD.18.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.19']['ZAD.19.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.20']['ZAD.20.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.21']['ZAD.21.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.22']['ZAD.22.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.23']['ZAD.23.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.24']['ZAD.24.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.25']['ZAD.25.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.26']['ZAD.26.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.27']['ZAD.27.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.28']['ZAD.28.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.29']['ZAD.29.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.30']['ZAD.30.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.31']['ZAD.31.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.32']['ZAD.32.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.33']['ZAD.33.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.34']['ZAD.34.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.35']['ZAD.35.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.36']['ZAD.36.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.37']['ZAD.37.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.38']['ZAD.38.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.39']['ZAD.39.1'].toString() == "")
                && (msg['ZAD'][count]['ZAD.40']['ZAD.40.1'].toString() == "")){
                                delete msg['ZAD'][count];
                
                }
                count++;
        }
    }
```
I'm getting the following error when trying to transform the sample message above 

Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	MIU DFT Out
CONNECTOR:	sourceConnector
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
102: 
103: 
104: var count=0;
105:     for each (segment in msg.children() ){
106:         if(segment.name() == "FT1"){
107:                 if ((msg['ZAD'][count]['ZAD.1']['ZAD.1.1'].toString() == "0") 
108:                 && (msg['ZAD'][count]['ZAD.2']['ZAD.2.1'].toString() == "") 
109:                 && (msg['ZAD'][count]['ZAD.3']['ZAD.3.1'].toString() == "")
110:                 && (msg['ZAD'][count]['ZAD.4']['ZAD.4.1'].toString() == "")
111:                 && (msg['ZAD'][count]['ZAD.5']['ZAD.5.1'].toString() == "")
LINE NUMBER:	107
DETAILS:	TypeError: Cannot read property "ZAD.1" from undefined
	at 85fad354-4983-40d7-a59e-86f9356537b1_JavaScript_Filter_Transformer_0:107 (doTransform)
	at 85fad354-4983-40d7-a59e-86f9356537b1_JavaScript_Filter_Transformer_0:174 (doScript)
	at 85fad354-4983-40d7-a59e-86f9356537b1_JavaScript_Filter_Transformer_0:176
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:235)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:187)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:114)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)

Does anyone have any thoughts or suggestions? I'm new to Mirth and I feel like the solution is sitting right in front of my nose. 

Thanks, 
Dom

**Comment by pacmano1** [comment]
@dombarone please format your code and use code blocks.    Read https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

**Comment by pacmano1** [comment]
Import @tonygermano xfilter code template from: https://github.com/nextgenhealthcare/connect-examples/tree/master/Code%20Templates/Filter%20XMLLists and make sure in channel dependencies it is avaiable to your channel.

Then:

```JS
function keepZadSegment(zadSegment) {
    // Check fields after ZAD|0|
    for (var i = 2; i < zadSegment.children().length(); i++) {
        if (zadSegment['ZAD.' + i].toString().trim() !== '') {
            return true; // Keep this segment as it has non-empty fields
        }
    }
    return false; // Remove this segment as all fields are empty
}

var filteredZads = xFilter(msg.ZAD, keepZadSegment, 0, Number.MAX_VALUE, false);

// Assign the filtered ZAD segments back to msg.ZAD
msg.ZAD = filteredZads;
```
Before:
```hl7
MSH|^~\&|||||19070618072145||ORU^R01|MSG0000000001|P|2.5|||ER
PID|1||1640666107||LUCAS^SANDRA^B|RICE^AUDREY^H||F|BRAME^BEVERLY^H
PV1|1|I
OBR|1
TQ1|1
OBX|1||||eirmod tempor invidunt ut labore et dolore||||||P~P
SPM|1|||||||||||||xxTESTxx||||19100213140054
ZAD|0||||
ZAD|0|||1|
```

After
```hl7
MSH|^~\&|||||19070618072145||ORU^R01|MSG0000000001|P|2.5|||ER
PID|1||1640666107||LUCAS^SANDRA^B|RICE^AUDREY^H||F|BRAME^BEVERLY^H
PV1|1|I
OBR|1
TQ1|1
OBX|1||||eirmod tempor invidunt ut labore et dolore||||||P~P
SPM|1|||||||||||||xxTESTxx||||19100213140054
ZAD|0|||1|
```

**Reply by pacmano1** [reply]
meh, maybe this is better, I was not checking for '0'

```JS
function keepZadSegment(zadSegment) {
    // If ZAD.1 is not '0', keep the segment
    if (zadSegment['ZAD.1']['ZAD.1.1'].toString().trim() != '0') {
        return true;
    }

    // If ZAD.1 is '0', check if any fields after ZAD|0| are non-empty
    for (var i = 2; i < zadSegment.children().length(); i++) {
        if (zadSegment['ZAD.' + i].toString().trim() != '') {
            return true; // Keep this segment as it has non-empty fields after ZAD|0|
        }
    }

    return false; // Remove this segment as ZAD.1 is '0' and all fields are empty after ZAD|0|
}

var filteredZads = xFilter(msg.ZAD, keepZadSegment, 0, Number.MAX_VALUE, false);

// Assign the filtered ZAD segments back to msg.ZAD
msg.ZAD = filteredZads;
```

**Reply by dombarone** [reply]
This got us most of the way there, The only thing I noticed is that If I have information in the second ZAD segment this seems to remove the first ZAD segment and then shift the second ZAD up in the message for example this:

MSH|^~\&|RCxRules Non-Appt|10020308|Ntierprise|Michigan Institute Of Urology PC|20231201095330||DFT^P03|8126f1a1-d4fd-448e-b987-83efb263e8e6|P|2.7|||AL
EVN|P03|20231201095330366
PID|1|11111113|11111113||TestIdaho^Duncan||19881127|M|||123 SPACE AVE ^^HARKONNEN^CA^67891||||||||999999999|||||||||||N
PV1|1|O|ASCHGOP|||||||||||||||||||||||||||||||||||||||||||||||
FT1|1|ae444048-94c1-4de9-86fb-0ea20546d181|1025|20231201^||CG|99213|||1|||MIU|||ASCHGOP|||I10^^I10|HOFF|JA01Anem||||99213|||||||
ZVA|ae444048-94c1-4de9-86fb-0ea20546d181
ZDS|||||||
ZAD|0|||||||||||||||||||||||||||||||||||||||
FT1|2|7d616e13-9845-4c24-a331-ada463d1f956|1025|20231201^||CG|80050|||1|||MIU|||ASCHGOP|||I10^^I10|HOFF|JA01Anem||||80050|||||||
ZVA|7d616e13-9845-4c24-a331-ada463d1f956
ZDS|||||||
ZAD|0||112|||||||||||||||||||||||||||||||||||||

Becomes this:

MSH|^~\&|RCxRules Non-Appt|10020308|Ntierprise|Michigan Institute Of Urology PC|20231201095330||DFT^P03|8126f1a1-d4fd-448e-b987-83efb263e8e6|P|2.7|||AL
EVN|P03|20231201095330366
PID|1|11111113|11111113||TestIdaho^Duncan||19881127|M|||123 SPACE AVE ^^HARKONNEN^CA^67891||||||||999999999|||||||||||N
PV1|1|O|ASCHGOP|||||||||||||||||||||||||||||||||||||||||||||||
FT1|1|ae444048-94c1-4de9-86fb-0ea20546d181|1025|20231201^||CG|99213|||1|||MIU|||ASCHGOP|||I10^^I10|HOFF|JA01Anem||||99213|||||||
ZVA|ae444048-94c1-4de9-86fb-0ea20546d181
ZDS|||||||
ZAD|0||112|||||||||||||||||||||||||||||||||||||
FT1|2|7d616e13-9845-4c24-a331-ada463d1f956|1025|20231201^||CG|80050|||1|||MIU|||ASCHGOP|||I10^^I10|HOFF|JA01Anem||||80050|||||||
ZVA|7d616e13-9845-4c24-a331-ada463d1f956
ZDS|||||||

Any thoughts on that? Thank you for getting me this far it is a HUGE improvement.

**Reply by pacmano1** [reply]
Oh, I thought the ZAD segments were all at the end.   I should have looked closer.  That code won't work.   One sec.

**Reply by pacmano1** [reply]
Are you receiving those messages?   Asking because a regex in the preprocessor can just strip those.

**Reply by pacmano1** [reply]
```JS
var hl7MessageString = message
var modifiedMessageString = hl7MessageString.replace(/^ZAD\|0(\|*)+\r?$/gm, '').replace(/\n\n/gm, '\n')
return modifiedMessageString;
```

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/0d57e951-a535-4278-a76a-74dc2285a32a)

Before:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/566ebb93-822c-4ed8-8459-c77547c22c80)

After regex:

![image](https://github.com/nextgenhealthcare/connect/assets/44065187/cce4d1d4-dc5a-43a6-b419-c9364caf44a8)

**Reply by dombarone** [reply]
The Regex seems to work but it is leaving an empty line in between the ZDS and the next FT1 are you doing anything special to remove that in your Mirth instance? I know my screenshot is in the raw tab but it looks the same in the sent message. 

![Screenshot 2023-12-01 153903](https://github.com/nextgenhealthcare/connect/assets/152644451/c5384a40-572d-420d-9c84-16e3690e84b3)

**Reply by pacmano1** [reply]
Do you have the second .replace ?   Show line ending on that screenshot also.   Post your preprocessor code snippet.

**Reply by dombarone** [reply]
![image](https://github.com/nextgenhealthcare/connect/assets/152644451/c09a83cd-cb92-4d09-9714-445757b522c6)
```javascript

var hl7MessageString = message
var modifiedMessageString = hl7MessageString.replace(/^ZAD\|0(\|*)+\r?$/gm, '').replace(/\n\n/gm, '\n')
return modifiedMessageString;
```

**Reply by pacmano1** [reply]
need to see the line endings, right click in the messge viewer and "Show Line Ending"

**Reply by dombarone** [reply]
![image](https://github.com/nextgenhealthcare/connect/assets/152644451/c4edde06-91df-4dbd-806e-2046257daa61)

**Reply by pacmano1** [reply, marked as answer]
you might need:

`var modifiedMessageString = hl7MessageString.replace(/^ZAD\|0(\|*)+\r?$/gm, '').replace(/\r\r/gm, '\r')`

I probably was pasting new lines for line ends, they should be carriage returns as per spec.

**Reply by dombarone** [reply]
Thank you!!! This is it. Have a great weekend.
